### ISSUE TITLE
Dynamic SSR per request

### ISSUE THREAD

**Issue by Celestialme** [sveltejs/kit#12941]
### Describe the problem

before i could enable/disable SSR per request from hooks, now its not possible. I don't really need SSR for my app but for SEO its better to be enabled.

### Describe the proposed solution

can we override export const ssr = false/true from hooks as it was before? so if I detect bot crawling I will return SSR true from hooks. 

### Importance

nice to have

**Comment by notramo** [comment]
I don't see any benefits of it. Why not enable SSR always?

**Comment by Celestialme** [comment]
because it flickers on hydration and is slow for me.

**Comment by Bishwas-py** [comment]
This could be a basic logic for this. But I am not exactly sure if `SSR` is handled by `preload` func or not.
```
const handleSSR: Handle = async ({ event, resolve }) => {
  const googleUserAgent = 'Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)';
  return resolve(event, {
    preload(input: { type: 'font' | 'css' | 'js' | 'asset'; path: string }): boolean {
      return event.request.headers.get('user-agent') !== googleUserAgent; // basic similar logic
    },
  });
};

export const handle = sequence(handleSSR);
```

Or, you can try create two separate servers, and using a middleman to where render the server with only SSR or not.

**Comment by dominikg** [comment]
just in case someone comes by here. google and potentially others are going to punish you if you present bots with a different experience than your regular users. Not sure if sending bots an ssr response and users not counts as that if the client rendering ultimately yields a similar experience, but don't come crying if you find your site downranked for playing dirty

**Comment by Celestialme** [comment]
> just in case someone comes by here. google and potentially others are going to punish you if you present bots with a different experience than your regular users. Not sure if sending bots an ssr response and users not counts as that if the client rendering ultimately yields a similar experience, but don't come crying if you find your site downranked for playing dirty

No it is called dynamic rendering and google is totally fine with that, I am doing this  about 6 years with previous svelte version and I am on first place. also you assume that SvelteKit SSR and CSR content is different and it does not work well enough? then it should be very noticeable for the eye when SSR to CSR happens in the browser. You are assuming this is cloaking which it is not. Cloaking is when you give bots different content which is not the case here.

for more to read
https://developers.google.com/search/docs/crawling-indexing/javascript/dynamic-rendering

yes it is says its not necessary now but not forbidden, but from my experience it is still superior to any rendering scheme.

**Comment by Celestialme** [comment]
> This could be a basic logic for this. But I am not exactly sure if `SSR` is handled by `preload` func or not.
> 
> ```ts
> const handleSSR: Handle = async ({ event, resolve }) => {
>   const googleUserAgent = 'Mozilla/5.0 (compatible; Googlebot/2.1; +http://www.google.com/bot.html)';
>   return resolve(event, {
>     preload(input: { type: 'font' | 'css' | 'js' | 'asset'; path: string }): boolean {
>       return event.request.headers.get('user-agent') !== googleUserAgent; // basic similar logic
>     },
>   });
> };
> 
> export const handle = sequence(handleSSR);
> ```
> 
> Or, you can try create two separate servers, and using a middleman to where render the server with only SSR or not.

yes creating separate build is totally possible but very inconvenient. Another approach is to create routes under SSR and CSR and use hooks reroute. but it is also very inconvenient.

**Comment by notramo** [comment]
Could you provide an example of how it flickers on hydration? Is it a theme change? Or the page reloads?

**Comment by Celestialme** [comment]
https://github.com/user-attachments/assets/3547785a-7c05-49dc-9bb9-c17afa581cc8


https://github.com/user-attachments/assets/dcf0d4a5-41e5-4405-8610-b897f565a959

on dev server you can see it flashes first flickers after. on build version not as bad just flashing.

**Comment by notramo** [comment]
This is not the expected outcome of an SSR. Svelte 5 has massively improved the hydration experience. Both the performance and features. My apps hydrate flawlessly and its UX is excellent.

I watched the video on 0.05× speed, and it became apparent that the SSR version only contains the header with the menu, but not the page main content. It's probably an issue with how your page is implemented.
The problem is that search engines will see the same empty page (only the header menu). Try disabling JS in your browser and reloading the page. That's what search engines see from your page. Tweak the page until the no-JS version shows the same content as the JS version. Obviously it won't be interactive, but since your page is static text and images, that content is achievable with noscript too.

Also, do you use something like `import { browser } from '$app/environment'`?

**Comment by Celestialme** [comment]
> This is not the expected outcome of an SSR. Svelte 5 has massively improved the hydration experience. Both the performance and features. My apps hydrate flawlessly and its UX is excellent.
> 
> I watched the video on 0.05× speed, and it became apparent that the SSR version only contains the header with the menu, but not the page main content. It's probably an issue with how your page is implemented. The problem is that search engines will see the same empty page (only the header menu). Try disabling JS in your browser and reloading the page. That's what search engines see from your page. Tweak the page until the no-JS version shows the same content as the JS version. Obviously it won't be interactive, but since your page is static text and images, that content is achievable with noscript too.
> 
> Also, do you use something like `import { browser } from '$app/environment'`?

thank you for looking into it. disabling javascript also renders full page. of course it is not interactable. that means page is rendered correctly on server side.

**Comment by Celestialme** [comment]
I found that in root +layout.svelte

`{#if browser || SSR}
	<slot />
{/if}
`
I can get SSR variable from load. now it does not flicker anymore and is way faster. but probably still nice to have more control from hooks side.

**Comment by mench32** [comment]
This is not a solution to the original problem.
This feature was present in previous versions of SvelteKit, and it’s surprising that it was removed.

Disabling SSR at runtime can significantly reduce server load.
For example, SSR could be disabled for authenticated users, which is completely safe for search engines, as they won’t see the content meant for authenticated users.

While it’s possible to come up with various workarounds and create different routes, it would be much more convenient to have the ability to control the SSR flag using the built-in tools of SvelteKit.

It would be great if this feature could be reinstated, as it would greatly improve the flexibility and performance of the framework. Thank you for your hard work and consideration!

**Comment by notramo** [comment]
>  For example, SSR could be disabled for authenticated users, which is completely safe for search engines, as they won’t see the content meant for authenticated users.

It's not that much of a gain, given that SSR is only utilized on the very first page load in the browser tab. After that, all navigations will use client-side changes, and only load `page.data` from the server.

**Comment by jippi** [comment]
I'm just chipping here that it would be very useful to have dynamic SSR based on per-request information (e.g. is bot or not)

Even if its a "first hit" only thing for users, when you have sufficiently many on them, running on low powered hardware, it can been as a significantly slower experience - _especially_ when you do a mix of DB and API lookup in routes depending on the user profile

having `export const ssr = my_function` that could access `getRequestEvent` and return a boolean would make it fairly straight forward, while retaining the current true/false option for folks who don't need it

Alternatively, if it could be controlled from a server hook, that would also work for me
